pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -954,18 +963,14 @@ protected void handleAuthResponse(CommandAuthResponse authResponse) {
         try {
             AuthData clientData = AuthData.of(authResponse.getResponse().getAuthData());
             doAuthentication(clientData, authResponse.getProtocolVersion(),
-                    authResponse.hasClientVersion() ? authResponse.getClientVersion() : EMPTY);
-        } catch (AuthenticationException e) {
-            service.getPulsarStats().recordConnectionCreateFail();
-            log.warn("[{}] Authentication failed: {} ", remoteAddress, e.getMessage());
-            ctx.writeAndFlush(Commands.newError(-1, ServerError.AuthenticationError, e.getMessage()));
-            close();
+                    authResponse.hasClientVersion() ? authResponse.getClientVersion() : EMPTY).whenComplete(
+                    (__, e) -> {
+                        if (e != null) {
+                            closeWithAuthException("authResponse", e);
+                        }
+                    });
         } catch (Exception e) {
-            service.getPulsarStats().recordConnectionCreateFail();
-            String msg = "Unable to handleAuthResponse";
-            log.warn("[{}] {} ", remoteAddress, msg, e);
-            ctx.writeAndFlush(Commands.newError(-1, ServerError.UnknownError, msg));
-            close();
+            closeWithAuthException("authResponse", e);
```

Comment:
The doAuthentication method will not throw exception directly right? It will complete the future with exception.

Reply:
Fixed.